### PR TITLE
Move deabstraction of non-tensorflow convention functions to the TFPartitioning pipeline.

### PR DESCRIPTION
**Pull request by bgogul**
This PR changes the invocation of deabstraction and partitioning passes as follows:

Diagnostic Pipeline: 
 - Deabstraction of _tensorflow-convention_ functions

Perf Pipeline: 
 - Partitioning of _tensorflow-convention_ functions (the next step is to move this to diagnostic pipeline so that we can have a proper -Onone mode)
 - Deabstraction and partitioning of _non-tensorflow convention_ functions. 

To recover most of the previous behavior with respect to sends and receives for non-tensorflow convention functions, the following optimizations are explicitly invoked after deabstraction (see `Deabtraction::doIt()`): 
```
  PredictableMemoryOptimizations();
  PropagateSSAValues()
```
Specifically, the TFPartitionPipeline looks like this now: 

```
SILPassPipelinePlan SILPassPipelinePlan::getTFPartitionPassPipeline() {
  SILPassPipelinePlan P;
  P.startPipeline("TensorFlow Partitioning");
  P.addTFDeabstraction();
  P.addTFPartition();
  return P;
}
```
Moreover, the perfInliner was modified to prevent inlining of some functions that the deabstraction relies on. 

Tests: 
 - Ignore the regressions in send/recv warnings in the tests as they are going away.
 - In some tests, the basic blocks got re-ordered because the deabstraction sees optimized SILFunction for non-tensorflow convention functions.

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Review by mhong — APPROVED** [review]
A very nice step forward!

**Comment by bgogul** [comment]
Please hold off further review on this PR. The changes that I did to the performance inliner (unsurprisingly) causes regressions in swift optimizer tests. I am looking into it.

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFDeabstraction.h
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.h]: Done.

File: lib/AST/TensorFlow.cpp
Comment (bgogul) [inline lib/AST/TensorFlow.cpp]: I will split this out into a separate CL.

**Review by bgogul — COMMENTED** [review]
@mhong, I have resolved all the comments and the optimization errors. PTAL.

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: can you provide an example on why we should care about optimizing such scalar (non-tensor) insts here? is it for tensorflow convention function?

Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: the larger question is that if we start applying more optimizations that used to belong only to the optimization pipeline, could that be "slippery slope" in that we might keep adding more such logic to the DA pass over time?

It'd be nice if we could try and make the DA pass compose with (and complimentary to) the other compiler passes as much as possible.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: can `deabstract` be a private method instead?

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp
Comment (mhong) [inline lib/SILOptimizer/Transforms/PerformanceInliner.cpp]: adding such domain specific (tensor specific) knowledge to this "general purpose" mechanism of performance inliner does not feel ideal. This might prevent us from being able to upstream this patch later.

is there a way to refactor the code so that the caller / constructor of SILInliner can pass in a callback (or a bit) to decide whether `func` is eligible for inlining? The call-site from TF can then pass in a callback which does the "tensor op check" as written here.

File: test/TensorFlow/control_flow.swift
Comment (mhong) [inline test/TensorFlow/control_flow.swift]: do you know why BB numbers are changing for some tests (consider adding the explanation to PR description)?

File: test/TensorFlow/crashers.swift
Comment (mhong) [inline test/TensorFlow/crashers.swift]: the changes in sends/recvs warnings are fairly distracting -- should i ignore all of them when reviewing this patch?

Ideally, we could send a patch to remove all such warnings first, as discussed.

File: test/TensorFlow/diagnostics.swift
Comment (mhong) [inline test/TensorFlow/diagnostics.swift]: tabs are used here and in some other places in this patch

**Review by mhong — COMMENTED** [review]
Left another around of comments. Thanks.

File: include/swift/AST/TensorFlow.h
Comment (bgogul) [reply include/swift/AST/TensorFlow.h]: No, it is not a cache. It keeps track of the nesting structure so that we can detect recursive data structures. 

In any case, ignore the file in this PR. I have opened a separate PR for this: https://github.com/apple/swift/pull/21449

(I had to include the files here to make sure I can run the tests.)

File: lib/AST/TensorFlow.cpp
Comment (bgogul) [reply lib/AST/TensorFlow.cpp]: https://github.com/apple/swift/pull/21449 has an example.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: This specific optimization mainly avoids the introduction of sends and receives by eliminating redundant allocations and creating an SSA value for attribute values.  This is actually quite similar to promoteToSSA function. (May be it is sufficient to call promoteToSSA again. Let me try it.)

When working on this PR, I simply tried to avoid introducing new sends and receives as much as possible in our tests. As a result of that approach, I identified that these optimizations were necessary. Given that we won't be running the optimization pipeline for tensor-flow convention functions in -Onone mode and still be able to extract graphs, we will have to perform some of the transformations here.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Done.

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp
Comment (bgogul) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp]: Yes, it is not ideal, but I prefer to leave this as is for now.

File: test/TensorFlow/control_flow.swift
Comment (bgogul) [reply test/TensorFlow/control_flow.swift]: I believe this has to do with the fact that the deabstraction sees optimized SIL function for non-tensorflow convention functions.

File: test/TensorFlow/crashers.swift
Comment (bgogul) [reply test/TensorFlow/crashers.swift]: Just ignore these for now. The good thing is that the runtime tests pass, which gives me *some* confidence that the refactoring is still generating correct code. 

(I will send out a separate PR removing these send/receive warnings along with adding an internal counter for debugging/testing purposes.)

File: test/TensorFlow/diagnostics.swift
Comment (bgogul) [reply test/TensorFlow/diagnostics.swift]: Oops...fixed now.

**Review by bgogul — COMMENTED** [review]
Investigating the failures in GPU and mac runs, but I have addressed all your comments otherwise. 

PTAL.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: done.

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp
Comment (mhong) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp]: I feel this is an important issue concering architecture ("layering") that we'll want to resolve ASAP. If you feel strongly about submitting this patch first, please go ahead, as long as we can follow up and address this one shortly.

File: test/TensorFlow/crashers.swift
Comment (mhong) [reply test/TensorFlow/crashers.swift]: How do we confirm that this patch has not overall regressed in the # of sends/recvs?

**Review by mhong — APPROVED** [review]
Left a few more comments, in terms of code behavior (verify no regression in sends/recvs) and readability (layering design). You can decide which ones to address before vs after submitting this patch.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: I have the same question as Mingsheng.  with partitioning running after the performance optimizer, it seems like these sorts of things should already be done.  Why is this necessary?  What is the bad thing that happens if we don't call into "optimizeMemoryAllocations" here?  Is this a phase ordering issue that can be resolved some other way?

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp
Comment (lattner) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Instead of matching on stage name, it would be cleaner to just have two passes, and insert them at the right part of hte pass pipeline.  The optimizer version of the partitioning pass should be a function pass, and the mandatory version should be a module pass, right?

File: lib/SILOptimizer/Transforms/PerformanceInliner.cpp
Comment (lattner) [reply lib/SILOptimizer/Transforms/PerformanceInliner.cpp]: I agree with Mingsheng that this is a horrible hack, but I'm ok with it to unblock work - so long as there is a known path out of this.  What is the plan to resolve this?

**Review by lattner — CHANGES_REQUESTED** [review]
I'd really like to understand why hard integration of optimizeMemoryAllocations is required. 

Please also split the deabstraction pass into two different passes (calling into a shared implementation).

Thanks!

**Review by bgogul — COMMENTED** [review]
Given our recent focus on tracing, some of the changes in this PR (e.g., related to optimizations and send/receives) could be deferred to later. However, I just wanted to reply to the comments and questions, when I have everything fresh in my memory. 

> I'd really like to understand why hard integration of optimizeMemoryAllocations is required.

Please see my reply to your review comment. 

> Please also split the deabstraction pass into two different passes (calling into a shared implementation).

I have factored out deabstraction as a utility and I am calling into it from partitioning now.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**
